cl/scrapers/tasks.py

```diff
@@ -228,6 +231,32 @@ def get_page_count(path, extension):
     return None
 
 
+def update_document_from_text(opinion):
+    """
+
+    :param opinion:
+    :return:
+    """
+    court = opinion.cluster.docket.court.pk
+    site = get_module_by_name(court)
+    metadata_dict = site.extract_from_text(opinion.plain_text)
+    models = apps.get_app_config("search").get_models()
+
+    while models:
+        mdl = models.pop()
+        if mdl.__name__ in metadata_dict.keys():
+            if mdl.__name__ == "OpinionCluster":
+                obj = mdl.objects.get(pk=opinion.cluster_id)
+                obj.__dict__.update(**metadata_dict[mdl.__name__])
+            elif mdl.__name__ == "Citation":
+                metadata_dict[mdl.__name__]["cluster_id"] = opinion.cluster_id
+                mdl.objects.get_or_create(**metadata_dict[mdl.__name__])
+            else:
+                raise NotImplemented(
+                    "Object type of %s not yet supported." % mdl.__name__
+                )
```

Comment:
OK, I don't know if the following will work, but I think it's a good simplification of what you've got. Why I like it:

1. The main benefit here is that we don't have two large variables we're working with. In your version, we're looking iterating over `models` and then looking stuff up in `metadata_dict` according to what we get in `models`. In my version, we eliminate `models`, iterate over `metadata_dict`, and do no lookups. 

2. It's shorter (at least until it's blackified, sigh)

3. I think it's more readable?

4. It gets the cluster in a much simpler way, which prevents a DB lookup. The cluster gets saved later on, so this should be fine, I think. 

```suggestion
for model_name, data in metadata_dict.items():
    ModelClass = apps.get_model("search.%s" % model_name)
    if model_name == "OpinionCluster":
        cluster = opinion.cluster.__dict__.update(data)    
    elif model_name == "Citation":
        ModelClass.objects.get_or_create(**data)
    else:
        raise NotImplemented("Object type of %s not yet supported." % model_name)
```

Anyway, this took me a few minutes of staring to figure out, and I haven't tested it, but it does feel more legible to me. (Your version, if we stick with it, would also benefit a lot by pulling `mdl.__name__` into a variable.)

Reply:
Does this need to be stored in the cluster 
```cluster = opinion.cluster.__dict__.update(data)  ```

Reply:
we still need to set the cluster id for citation generation
```data["cluster_id"] = op.cluster_id```

Reply:
I think this works
```     
for model_name, data in cd.items():
      ModelClass = apps.get_model("search.%s" % model_name)
      if model_name == "OpinionCluster":
          op.cluster.__dict__.update(**data)
      elif model_name == "Citation":
          data["cluster_id"] = op.cluster_id
          ModelClass.objects.get_or_create(**data)
      else:
          raise NotImplemented("Object type of %s not yet supported." % model_name```

Reply:
I think this has been resolved